bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_3drepo.cpp

```diff
@@ -132,25 +137,76 @@ repo::core::model::MaterialNode *RepoModelImport::parseMaterial(const boost::pro
 	materials.insert(materialNode);
 	matNodeList.push_back(materialNode);
 
-	return materialNode;
+	if(textureId >= 0)
+	{
+		textureIdToParents[textureId].push_back(materialNode->getSharedID());
+	}
+	return;
+}
+
+void RepoModelImport::parseTexture(
+	const boost::property_tree::ptree& textureTree, 
+	char * const dataBuffer)
+{
+	bool fileNameOk	 = textureTree.find("filename") != textureTree.not_found();
+	bool byteCountOK = textureTree.find("numImageBytes") != textureTree.not_found();
+	bool widthOk	 = textureTree.find("width") != textureTree.not_found();
+	bool heightOk	 = textureTree.find("height") != textureTree.not_found();
+	bool idOk		 = textureTree.find("id") != textureTree.not_found();
+
+	if (!byteCountOK ||
+		!widthOk	 ||
+		!heightOk	 ||
+		!idOk)
+	{
+		repoError << "Required texture field missing. Skipping this texture.";
+		missingTextures = true;
+		return;
+	}
+
+	std::string name = fileNameOk ? textureTree.get_child("filename").data() : "";
+	uint32_t byteCount = textureTree.get<uint32_t>("numImageBytes");
+	uint32_t width = textureTree.get<uint32_t>("width");
+	uint32_t height = textureTree.get<uint32_t>("height");
+	uint32_t id = textureTree.get<uint32_t>("id");
+
+	std::vector<uint32_t> DataStartEnd = as_vector<uint32_t>(textureTree, "imageBytes");
+
+	char* data = &dataBuffer[DataStartEnd[0]];
+
+	repo::core::model::TextureNode* textureNode =
+		new repo::core::model::TextureNode(
+			repo::core::model::RepoBSONFactory::makeTextureNode(
+				name,
+				data,
+				byteCount,
+				width,
+				height,
+				REPO_NODE_API_LEVEL_1));
+	
+	repo::core::model::TextureNode* tmpTexture = new repo::core::model::TextureNode();
+	*tmpTexture = textureNode->cloneAndAddParent(textureIdToParents[id]);
```

Comment:
alter the `makeTextureNode` api to optionally take in parents array. cloning is expensive so we want to avoid doing that unless it's necessary

Reply:
I have updated the method to do this here:
https://github.com/3drepo/3drepobouncer/blob/b18deadb51eec7122568223e98e775d5a145a670/bouncer/src/repo/core/model/bson/repo_bson_factory.cpp#L939

this problem still exists for material nodes that still use the clone method but the fix for this would require geometry to be parsed before materials, I made a ticket for this [here ](https://github.com/3drepo/3drepobouncer/issues/416)